### ISSUE TITLE
Incompatible with Multiplayer

### ISSUE THREAD

**Issue by DiscipleOfEris** [modo-lv/factorio-mod-stack-combinator#39]
When players attempt to join a multiplayer server with the mod enabled, they get this error:

![image](https://user-images.githubusercontent.com/45589381/128935893-eddc2259-06d8-49f5-8869-294d725ec6d7.png)

I've attached the log, but the important lines are 94-96:
```
11.934 Error ClientMultiplayerManager.cpp:1114: mod-stack-combinator was not registered for the following events when the map was saved but has registered them as a result of loading: on_tick (ID 0)
11.934 Error ClientMultiplayerManager.cpp:1130: mod-stack-combinator was registered for the following nth_ticks when the map was saved but has not re-registered them as a result of loading: 1
11.934 Error ClientMultiplayerManager.cpp:98: MultiplayerManager failed: "" + multiplayer.script-event-mismatch + "
```

[factorio-current.log](https://github.com/modo-lv/factorio-mod-stack-combinator/files/6964590/factorio-current.log)

Steps to reproduce:
1. Launch game (v1.1.37) through Steam. Disable all mods. Enable Stack Combinator (v1.2.8).
2. Host a multiplayer game from a new save. Doesn't matter if the game has any StaCo's in it.
3. Attempt to join with a second player. (Sync Mods if necessary).

Steps to test multiplayer by yourself:
1. Log into Factorio.com. Download the .zip of the full game. Unzip wherever.
2. Launch game through Steam. Host save as before but with "Verify User Identity" disabled.
3. Launch game through downloaded Factorio. You either need to launch with arguments to point at your mod folder, or login to download the mod, then go into Settings/Other to logout and change name to "Test" or something.
4. Attempt to join LAN game.

**Comment by modo-lv** [comment]
Ah, this must have broken in the last update where I added the update delay setting. I'll try to fix it ASAP, but in the meantime the older version (1.2.7) should work.

**Comment by DiscipleOfEris** [comment]
When I attempted to join a brand new save with only the Stack Combinator (v1.2.7) mod enabled (on both players) I got this:

![image](https://user-images.githubusercontent.com/45589381/129075361-60680a9d-7548-4995-8510-abeddc391e49.png)

The desync report is uploaded, but here are what seem to be the important lines from the log:
```
  28.408 Checksum for script C:/Users/ofsha/Downloads/Factorio_1.1.37/temp/currently-playing/control.lua: 694211064
  28.417 Checksum for script __stack-combinator__/control.lua: 383861242
  28.429 Info ClientMultiplayerManager.cpp:592: UpdateTick(2217) changing state from(ConnectedLoadingMap) to(TryingToCatchUp)
  28.449 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2218) serverCRC(236483267) localCRC(3297570422)
  28.449 Info ClientMultiplayerManager.cpp:592: UpdateTick(2219) changing state from(TryingToCatchUp) to(DesyncedWaitingForMap)
  28.461 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2219) serverCRC(1329794167) localCRC(2245497026)
  28.479 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2220) serverCRC(2030102585) localCRC(3012850828)
  28.495 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2221) serverCRC(1401733094) localCRC(2568495955)
  28.511 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2222) serverCRC(677612685) localCRC(3807842360)
  28.529 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2223) serverCRC(1565268986) localCRC(2547550031)
  28.545 Error GameActionHandler.cpp:3557: Multiplayer desynchronisation: crc test (heuristic) failed for crcTick(2224) serverCRC(1844615784) localCRC(2808481501)
  28.561 Info ClientMultiplayerManager.cpp:762: updateTick(2226) received player (1) desynced
```

[desync-report-2021-08-11_10-26-45.zip](https://github.com/modo-lv/factorio-mod-stack-combinator/files/6970277/desync-report-2021-08-11_10-26-45.zip)

Perhaps the event handlers need to be registered in on_load or they're not registered at the same time on both players, causing the desync. I haven't really inspected the code enough to offer much more feedback than that.

**Comment by modo-lv** [comment]
Hm, that's new. I'll try to free up some time over the weekend for this.

**Comment by modo-lv** [comment]
I have published version 1.2.9, which fixes several desync issues. I can't guarantee it will work in existing MP saves, but new games should work fine. Please try it out and let me know if anything's still not working right!

**Comment by DiscipleOfEris** [comment]
I'm not sure what I did wrong, but I ran a new singleplayer Sandbox game with nothing but Stack Combinator 1.2.9 (and dependencies) enabled. The mod crashes as soon as I select circuit wire. I can host and join multiplayer games and place StaCos and don't get any desync problems, but again, as soon as any player selects a circuit wire (green wire) the mod crashes, even if the map is completely empty. I can load a Freeplay map, but again the mod crashes if green wire is selected.

![image](https://user-images.githubusercontent.com/45589381/129612095-442b089b-c127-4539-b829-99d1381d83bc.png)

**Comment by modo-lv** [comment]
This was a new side-bug on my part due to all the multiplayer fixing. I've just fixed it and published v1.2.10.

**Comment by DiscipleOfEris** [comment]
Alright, after my testing it seems to be working fine in old and new saves in single- and multiplayer. I did not have a save that already contained a StaCo to test with, however, so all my tests were fresh, at least w.r.t. Stack Combinator.
